File: app/vortasks/lib/models/goals/goal_history.dart
import 'package:intl/intl.dart';
import 'package:vortasks/models/enums/goal_type.dart';
import 'package:vortasks/models/enums/task_type.dart';

class GoalHistory {
  final String id;
  final DateTime date;
  final int weekNumber;
  final TaskType type;
  final GoalType goalType;
  int successes;
  int failures;

  GoalHistory({
    required this.id,
    required this.date,
    required this.weekNumber,
    required this.type,
    required this.goalType,
    required this.successes,
    required this.failures,
  });

  factory GoalHistory.fromJson(Map<String, dynamic> json) {
    return GoalHistory(
      id: json['id'],
      date: DateTime.parse(json['date']),
      weekNumber: json['weekNumber'],
      type: TaskType.values.byName(json['type']),
      goalType: GoalType.values.byName(json['goalType']),
      successes: json['successes'],
      failures: json['failures'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'date': DateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'").format(date),
      'weekNumber': weekNumber,
      'type': type.name,
      'goalType': goalType.name,
      'successes': successes,
      'failures': failures,
    };
  }
}

File: api/src/main/java/com/leandroadal/vortasks/entities/backup/userprogress/GoalHistory.java
package com.leandroadal.vortasks.entities.backup.userprogress;

import java.time.Instant;

import com.leandroadal.vortasks.entities.backup.Backup;
import com.leandroadal.vortasks.entities.backup.userprogress.enums.GoalType;
import com.leandroadal.vortasks.entities.backup.userprogress.enums.TaskType;

import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Getter
@Setter
@Entity
@NoArgsConstructor
public class GoalHistory {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private String id;

    private Instant date;
    private Integer weekNumber;

    @Enumerated(EnumType.STRING)
    private TaskType type;

    @Enumerated(EnumType.STRING)
    private GoalType goalType;

    private Integer successes;
    private Integer failures;

    @ManyToOne
    @JoinColumn(name = "backup_id")
    private Backup backup;
}
